From 70bd02f2ea3a91c0340df97ca3b8a4f9400f6f46 Mon Sep 17 00:00:00 2001
From: Peter Hutterer <peter.hutterer@who-t.net>
Date: Tue, 23 Mar 2010 12:13:56 +1000
Subject: [PATCH 29/37] xfree86: merge driver from the input class into the options.

A driver that is assigned by an input class is only present as idev->driver.
The driver itself has no access to this information once PreInit is called.
For devices that rely on chain-hotplugging (wacom), this means that for the
second device the driver information is lost and the second device cannot be
initialized through NewInputDeviceRequest. Although this could be worked
around by hardcoding the driver name in the wacom driver, having the
assigned driver in the options seems like the better solution.

This issue only manifests itself with the udev backend. With HAL, the driver
is assigned by HAL and the option is duplicated in config/hal.c.

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>
Reviewed-by: Dan Nicholson <dbn.lists@gmail.com>
---
 hw/xfree86/common/xf86Xinput.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git xorg-server-1.7.99.902/hw/xfree86/common/xf86Xinput.c xorg-server-1.7.99.902/hw/xfree86/common/xf86Xinput.c
index e9f7f2f..8229227 100644
--- xorg-server-1.7.99.902/hw/xfree86/common/xf86Xinput.c
+++ xorg-server-1.7.99.902/hw/xfree86/common/xf86Xinput.c
@@ -618,6 +618,7 @@ MergeInputClasses(IDevPtr idev, InputAttributes *attrs)
                     "InputClass configuration");
             return BadAlloc;
         }
+        mergedopts = xf86ReplaceStrOption(mergedopts, "driver", idev->driver);
     }
     idev->commonOptions = xf86optionListMerge(idev->commonOptions, mergedopts);
 
-- 
1.7.0.2

